### PR TITLE
1009 immunization v2 update the patient profile to only should due immunization tasks

### PR DESCRIPTION
**Pull request by ahsanbhatti49**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes #[issue number] or Closes #[issue number]

**Engineer Checklist**
- [ ] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [ ] I have added any strings visible on UI components to the `strings.xml` file
- [ ] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [ ] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [ ] I have built and run the FHIRCore app to verify my change fixes the issue and/or does not break the app 

**Code Reviewer Checklist**
- [ ] I have verified **Unit tests** have been written for any new feature(s) and edge cases
- [ ] I have verified any strings visible on UI components are in the `strings.xml` file
- [ ] I have verifed the [CHANGELOG.md](./CHANGELOG.md) file has any notable changes to the codebase
- [ ] I have verified the solution has been implemented in a configurable and generic way for reuseable components
- [ ] I have built and run the FHIRCore app to verify the change fixes the issue and/or does not break the app

### CODE REVIEW

**Comment by ahsanbhatti49** [comment]
This ticket refers to [https://github.com/opensrp/fhir-resources/issues/1009](url)

File: android/engine/src/test/resources/extractions/record-all-immunization/structure-map.txt [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+                 ExtractImmunization(src, encounter, tgt) "r_careplan";
+}
+
+group ExtractEncounter(source src : QuestionnaireResponse, target encounter: Encounter, target bundle : Bundle) {
+        src -> encounter.id = uuid(),
+            encounter.status = "finished",
+            encounter.serviceType = cc("http://terminology.hl7.org/CodeSystem/service-type", "57", "Immunization") "rule_ee_4";
+
+        src -> bundle.entry as entry, entry.resource = encounter "r_enc_cp";
+}
```
Comment (dubdabasoduba) [inline android/engine/src/test/resources/extractions/record-all-immunization/structure-map.txt:18]: If this is to fulfil the following discussion then the encounters generated are not correct. https://github.com/opensrp/fhircore/discussions/2142 We need 2 Encounters + 1 Immunization resources

Comment (dubdabasoduba) [reply android/engine/src/test/resources/extractions/record-all-immunization/structure-map.txt:18]: Also, the encounters do not have all the required fields.

File: android/quest/src/main/assets/configs/app/profiles/default_profile_config.json [hunk 2]
```diff
@@ -1231,7 +1231,7 @@
                   "id": "availableTasks",
                   "relatedResourceId": "availableTasks",
                   "resourceType": "Task",
-                  "conditionalFhirPathExpression": "Task.reasonCode.coding[0].code = 'immunization_at_birth'"
+                  "conditionalFhirPathExpression": "Task.groupIdentifier.value = '0_d' and Task.executionPeriod.start > today() + 7 days"
```
Comment (dubdabasoduba) [inline android/quest/src/main/assets/configs/app/profiles/default_profile_config.json:1234]: I think this timings should be `today() - 7days`. We want to show the tasks 7 days before they are DUE.

File: android/quest/src/main/assets/configs/app/profiles/default_profile_config.json [hunk 3]
```diff
@@ -1351,7 +1351,7 @@
                   "id": "availableTasks",
                   "relatedResourceId": "availableTasks",
                   "resourceType": "Task",
-                  "conditionalFhirPathExpression": "Task.reasonCode.coding[0].code = 'immunization_at_6_weeks'"
+                  "conditionalFhirPathExpression": "Task.groupIdentifier.value = '6_wk' and Task.executionPeriod.start > today() + 7 days"
```
Comment (dubdabasoduba) [inline android/quest/src/main/assets/configs/app/profiles/default_profile_config.json:1354]: Same as the timing comment

File: android/quest/src/main/assets/configs/app/profiles/default_profile_config.json [hunk 4]
```diff
@@ -1471,7 +1471,7 @@
                   "id": "availableTasks",
                   "relatedResourceId": "availableTasks",
                   "resourceType": "Task",
-                  "conditionalFhirPathExpression": "Task.reasonCode.coding[0].code = 'immunization_at_10_weeks'"
+                  "conditionalFhirPathExpression": "Task.groupIdentifier.value = '10_wk' and Task.executionPeriod.start > today() + 7 days"
```
Comment (dubdabasoduba) [inline android/quest/src/main/assets/configs/app/profiles/default_profile_config.json:1474]: Same as the timing comment

File: android/quest/src/main/assets/configs/app/profiles/default_profile_config.json [hunk 5]
```diff
@@ -1591,7 +1591,7 @@
                   "id": "availableTasks",
                   "relatedResourceId": "availableTasks",
                   "resourceType": "Task",
-                  "conditionalFhirPathExpression": "Task.reasonCode.coding[0].code = 'immunization_at_14_weeks'"
+                  "conditionalFhirPathExpression": "Task.groupIdentifier.value = '14_wk' and Task.executionPeriod.start > today() + 7 days"
```
Comment (dubdabasoduba) [inline android/quest/src/main/assets/configs/app/profiles/default_profile_config.json:1594]: Same as the timing comment

File: android/quest/src/main/assets/configs/app/profiles/default_profile_config.json [hunk 6]
```diff
@@ -1711,7 +1711,7 @@
                   "id": "availableTasks",
                   "relatedResourceId": "availableTasks",
                   "resourceType": "Task",
-                  "conditionalFhirPathExpression": "Task.reasonCode.coding[0].code = 'immunization_at_9_months'"
+                  "conditionalFhirPathExpression": "Task.groupIdentifier.value = '9_mo' and Task.executionPeriod.start > today() + 7 days"
```
Comment (dubdabasoduba) [inline android/quest/src/main/assets/configs/app/profiles/default_profile_config.json:1714]: Same as the timing comment

File: android/quest/src/main/assets/configs/app/profiles/default_profile_config.json [hunk 7]
```diff
@@ -1831,7 +1831,7 @@
                   "id": "availableTasks",
                   "relatedResourceId": "availableTasks",
                   "resourceType": "Task",
-                  "conditionalFhirPathExpression": "Task.reasonCode.coding[0].code = 'immunization_at_15_months'"
+                  "conditionalFhirPathExpression": "Task.groupIdentifier.value = '15_mo' and Task.executionPeriod.start > today() + 7 days"
```
Comment (dubdabasoduba) [inline android/quest/src/main/assets/configs/app/profiles/default_profile_config.json:1834]: Same as the timing comment

File: android/quest/src/main/assets/configs/app/profiles/default_profile_config.json [hunk 8]
```diff
@@ -1951,7 +1951,7 @@
                   "id": "availableTasks",
                   "relatedResourceId": "availableTasks",
                   "resourceType": "Task",
-                  "conditionalFhirPathExpression": "Task.reasonCode.coding[0].code = 'immunization_at_9_years'"
+                  "conditionalFhirPathExpression": "Task.groupIdentifier.value = '108_mo' and Task.executionPeriod.start > today() + 7 days"
```
Comment (dubdabasoduba) [inline android/quest/src/main/assets/configs/app/profiles/default_profile_config.json:1954]: Same as the timing comment

File: android/quest/src/main/assets/configs/app/profiles/default_profile_config.json [hunk 9]
```diff
@@ -2071,7 +2071,7 @@
                   "id": "availableTasks",
                   "relatedResourceId": "availableTasks",
                   "resourceType": "Task",
-                  "conditionalFhirPathExpression": "Task.reasonCode.coding[0].code = 'immunization_at_9_half_years'"
+                  "conditionalFhirPathExpression": "Task.groupIdentifier.value = '114_wk' and Task.executionPeriod.start > today() + 7 days"
```
Comment (dubdabasoduba) [inline android/quest/src/main/assets/configs/app/profiles/default_profile_config.json:2074]: Same as the timing comment

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]

File: android/engine/src/test/resources/extractions/record-all-immunization/structure-map.txt (hunk 1 above)
Comment (maimoonak) [reply android/engine/src/test/resources/extractions/record-all-immunization/structure-map.txt:18]: @dubdabasoduba This was to test and add a base for extraction in record all and was done as part of https://github.com/opensrp/fhir-resources/issues/1006 ticket. We can add corresponding entities as per requirement. Kindly share the ticket number if any is specific for this. This looks like FhirCareplanGenerateTest but it is specifically to add / test extraction of questionnaires which is different from CarePlan generation.

File: android/engine/src/test/java/org/smartregister/fhircore/engine/FhirExtractionTest.kt
[the remark this answers is not in the document]
Comment (maimoonak) [reply android/engine/src/test/java/org/smartregister/fhircore/engine/FhirExtractionTest.kt:1]: The class was copied from FhirCareplanGenerateTest which was modified later. There was an issue with branches merging which ended up the updates missed. Would be updated shortly

File: android/engine/src/test/java/org/smartregister/fhircore/engine/FhirExtractionTest.kt [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    }
+  }
+  data class ExtractionResources(
+    val questionnaire: Questionnaire,
+    val patient: Patient,
+    val questionnaireResponse: QuestionnaireResponse,
+    val structureMap: StructureMap,
+    val resourcesSlot: MutableList<Resource>
+  )
+  fun loadExtractionResources(name: String): ExtractionResources {
```
Comment (pld) [inline android/engine/src/test/java/org/smartregister/fhircore/engine/FhirExtractionTest.kt:133]:
```suggestion

  @Before
  fun setup() {
    hiltRule.inject()
    structureMapUtilities = StructureMapUtilities(transformSupportServices.simpleWorkerContext)
    val workManager = mockk<WorkManager>()
    every { defaultRepository.dispatcherProvider.io() } returns Dispatchers.IO
    every { defaultRepository.fhirEngine } returns fhirEngine
    every { workManager.enqueue(any<WorkRequest>()) } returns mockk()
  }

  @Test
  @ExperimentalCoroutinesApi
  fun `extract should generate immunization and encounter`() = runTest {
    val resources = loadExtractionResources("record-all-immunization")
    val questionnaire = resources.questionnaire
    val patient = resources.patient.apply { birthDate = Date() }
    val questionnaireResponse = resources.questionnaireResponse
    val result =
      ResourceMapper.extract(
          questionnaire = questionnaire,
          questionnaireResponse = questionnaireResponse,
          structureMapExtractionContext =
            StructureMapExtractionContext(
              context = context,
              structureMapProvider = { _, _ -> resources.structureMap },
              transformSupportServices = transformSupportServices
            )
        )
        .also { println(it.encodeResourceToString()) }
    val encounter = result.entry.find { it.resource is Encounter }!!.resource as Encounter
    result.entry.filter { it.resource is Immunization }.map { it.resource as Immunization }.also {
      assertEquals(3, it.size)
      assertTrue(it.all { it.encounter.reference == encounter.referenceValue() })
      assertTrue(it.all { it.status == Immunization.ImmunizationStatus.COMPLETED })
      assertTrue(it.all { it.recorded.makeItReadable() == Date().makeItReadable() })
      val bcg = it[0]
      assertEquals(bcg.occurrenceDateTimeType.value.makeItReadable(), "14-Apr-2023")
      assertEquals(bcg.vaccineCode.text, "BCG")
      assertEquals(bcg.vaccineCode.codingFirstRep.code, "42284007")
      assertEquals(bcg.vaccineCode.codingFirstRep.display, "BCG vaccine")
      assertEquals(bcg.vaccineCode.codingFirstRep.system, "http://snomed.info/sct")
      val opv1 = it[1]
      assertEquals(opv1.occurrenceDateTimeType.value.makeItReadable(), "21-Apr-2023")
      assertEquals(opv1.vaccineCode.text, "OPV 1")
      assertEquals(opv1.vaccineCode.codingFirstRep.code, "111164008")
      assertEquals(opv1.vaccineCode.codingFirstRep.display, "Poliovirus vaccine")
      assertEquals(opv1.vaccineCode.codingFirstRep.system, "http://snomed.info/sct")
      val rota1 = it[2]
      assertEquals(rota1.occurrenceDateTimeType.value.makeItReadable(), "28-Apr-2023")
      assertEquals(rota1.vaccineCode.text, "ROTA 1")
      assertEquals(rota1.vaccineCode.codingFirstRep.code, "415354003")
      assertEquals(rota1.vaccineCode.codingFirstRep.display, "Rotavirus vaccine")
      assertEquals(rota1.vaccineCode.codingFirstRep.system, "http://snomed.info/sct")
    }
  }

  data class ExtractionResources(
    val questionnaire: Questionnaire,
    val patient: Patient,
    val questionnaireResponse: QuestionnaireResponse,
    val structureMap: StructureMap,
    val resourcesSlot: MutableList<Resource>
  )

  fun loadExtractionResources(name: String): ExtractionResources {
```

**Review by pld — APPROVED** [review]

**Pull request merged**